### PR TITLE
feat: add DeepSeek V4 Flash 0731 renderer

### PR DESCRIPTION
**Pull request by hallerite**
## Summary

- add a hand-coded DeepSeek V4 renderer matching the official Python encoding_dsv4.py contract
- support chat and thinking modes, reasoning effort, DSML tool schemas and calls, ordered tool results, parsing, attribution masks, and safe bridging
- register deepseek-ai/DeepSeek-V4-Flash-0731 under the typed deepseek-v4 configuration and document the new renderer
- generalize the shared test oracle so Jinja checkpoints use apply_chat_template while DeepSeek V4 uses an independent mirror of its official Python encoder
- add DeepSeek V4 to the shared render, attribution, parsing, preservation, bridge, round-trip, and config-parity matrices
- encode contiguous prompt text in one BPE pass; the generalized barrage caught token-ID drift at renderer chunk boundaries even when decoded strings matched
- preserve the official encoder's task-transition, user/tool merge, and strict DSML parsing edge behavior with explicit parity regressions
- merge the latest origin/main (32e6622) with a signed merge commit after re-signing the full PR stack

## Validation

- all 4 official DeepSeek encoding fixtures match byte-for-byte
- DeepSeek V4 shared barrage: 68 passed, 1 expected skip
- DeepSeek-focused bridge, round-trip, dedicated, and config matrix: 71 passed, 1 expected skip
- exact task/DSML parity regressions: 19 passed
- official JSON non-object argument exception parity: 4 passed
- focused DeepSeek V4 file after final parity regression: 23 passed
- post-refactor parser and round-trip verification: 36 passed
- full merged-tree suite: 3821 passed, 181 skipped, 1 expected xfail
- repository-wide Ruff lint and format checks pass
- CI-equivalent renderer type check completes without the parser panic exposed during development

Official encoder: https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-0731/blob/main/encoding/encoding_dsv4.py

> [!NOTE]
> ### Add `DeepSeekV4Renderer` for DeepSeek V4 Flash 0731
> - Adds `DeepSeekV4Renderer` and `DeepSeekV4RendererConfig` to support the `deepseek-ai/DeepSeek-V4-Flash-0731` model, which uses DSML wire format instead of a tokenizer chat template.
> - Implements DSML tool rendering, tool-result merging and ordering, thinking retention logic, task-token emission, and response-format embedding in [deepseek_v4.py](https://github.com/PrimeIntellect-ai/renderers/pull/140/files#diff-2b0e8e3fed0ce308b759b25c1dc30d12c73b6ef9be7fcf3442758a19d2434503).
> - Adds `parse_deepseek_v4` to decode model outputs into content, reasoning, and structured DSML tool calls in [parsing.py](https://github.com/PrimeIntellect-ai/renderers/pull/140/files#diff-4eba1ee111275cd49ea938c89bfbb7dafa7894d18a97c858fb4a9df401e85c75).
> - Introduces a model-aware reference oracle in [reference_rendering.py](https://github.com/PrimeIntellect-ai/renderers/pull/140/files#diff-c78d8e4e4db6f77a32a16213567941e60cc106efc3e165106be57789914dbbfd) to replace `tokenizer.apply_chat_template` in test parity checks.
> - Behavioral Change: `DeepSeekV4RendererConfig` defaults to `enable_thinking=False`, `drop_thinking=True`, and `reasoning_effort='low'`.
>
> 
> #### Changes since #140 opened
>
> - Introduced `_QUERY_ROLES` frozenset constant containing 'user' and 'developer', added `_is_query_message` helper function, and updated `renderers.deepseek_v4.DeepSeekV4Renderer.render_ids` to replace hardcoded role checks with `_QUERY_ROLES` when computing the last query boundary index, filtering historical developer and search-agent messages during thinking drops, determining reasoning retention for assistant messages, and opening new reasoning segments [98cb612]
> - Updated `renderers.deepseek_v4.DeepSeekV4Renderer.bridge_to_next_turn` to pass `is_user_query=_is_query_message` predicate to `should_rerender_for_thinking_retention` method [98cb612]
> - Added test cases `test_bridge_declines_at_developer_query_boundary_when_dropping_thinking` and `test_bridge_extends_developer_query_when_preserving_all_thinking` in `tests.test_deepseek_v4` module [98cb612]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized a7217e5.</sup>
> 
>

---

> [!NOTE]
> **Medium Risk**
> New ~860-line renderer on the critical prompt/tokenization path with complex thinking, tool, and bridge semantics; mistakes would skew training/inference token streams for V4, though coverage is extensive and defaults match the official encoder.
> 
> **Overview**
> Adds first-class support for **`deepseek-ai/DeepSeek-V4-Flash-0731`**, which has no Jinja chat template and instead follows the model repo’s Python **`encoding_dsv4.py`** contract.
> 
> **`DeepSeekV4Renderer`** and **`DeepSeekV4RendererConfig`** implement that encoder in the renderer protocol: chat vs thinking mode, optional reasoning-effort prefixes, historical reasoning drop (disabled when tools are present), DSML tool schemas/calls, merged **`tool`** results with call-order sorting, task tokens, response-format blocks, **`parse_deepseek_v4`** for completions, and **`bridge_to_next_turn`** with thinking-retention rules. Prompt text is tokenized in **one BPE pass** so token IDs match the reference encoder across fragment boundaries.
> 
> Registration wires the checkpoint to **`deepseek-v4`** via **`MODEL_RENDERER_MAP`**, lazy exports, and the typed config union; README and **`docs/renderer-config.md`** document the new family.
> 
> Tests add **`tests/reference_rendering.render_reference`** so shared parity uses **`apply_chat_template`** for Jinja models and an independent V4 mirror for barrage/bridge/config-matrix coverage, plus **`tests/test_deepseek_v4.py`** for encoder edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 98cb612bb83c5f782668e25ee5da17c6550795d5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: renderers/deepseek_v4.py [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+                    continue
+                emit_special(self._assistant, transition_msg_idx)
+                emit_special(
+                    self._think_start
+                    if self.config.enable_thinking
+                    else self._think_end,
+                    transition_msg_idx,
+                )
+                emit_special(self._task_ids[task], transition_msg_idx)
+                continue
```
[the remark this answers is not in the document]
Comment (hallerite) [reply renderers/deepseek_v4.py:745]: Confirmed against the official Python encoder. Action tasks select the thinking opener from thinking_mode alone, and an assistant immediately following any task suppresses all reasoning markup. Commit fba404e now matches that behavior in both current and historical action turns, and adds direct regression coverage. It also fixes the adjacent real divergence where our current post-task assistant previously emitted a closing thinking token that upstream omits.

File: renderers/deepseek_v4.py [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+            continue
+
+        if role == "user":
+            block = _ContentBlock(
+                kind="text",
+                content=_text_content(message.get("content")),
+                message_index=index,
+            )
+            if merged and merged[-1].role == "user" and merged[-1].task is None:
+                merged[-1].blocks.append(block)
```
[the remark this answers is not in the document]
Comment (hallerite) [reply renderers/deepseek_v4.py:242]: Confirmed exact upstream parity. DeepSeek merge_tool_messages checks only whether the previous user turn has a task, then appends the incoming text block without copying the incoming task. We intentionally retain that behavior. Commit fba404e adds a direct token-parity regression for this merge case.

File: renderers/parsing.py [hunk 3]
```diff
@@ -1149,6 +1149,202 @@ def _parse_deepseek_tool_calls(
...
+    content_ids = ids[content_offset:]
+    decoded = _decode(tokenizer, content_ids)
+    section_marker = "\n\n<｜DSML｜tool_calls>"
+    section_pos = decoded.find(section_marker)
+    if section_pos == -1 or dsml_id not in content_ids:
+        return ParsedResponse(
+            content=decoded,
+            reasoning_content=reasoning or None,
+            tool_calls=[],
+        )
```
[the remark this answers is not in the document]
Comment (hallerite) [reply renderers/parsing.py:1196]: Confirmed exact upstream parser parity. DeepSeek parse_message_from_completion_text uses the same exact two-newline DSML sentinel, and the canonical encoder always emits that form. We intentionally do not accept zero or one newline here. Commit fba404e adds parameterized regression coverage pinning those variants as plain content.

File: renderers/deepseek_v4.py [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+                )
+
+            next_role = (
+                logical_messages[index + 1].role
+                if index + 1 < len(logical_messages)
+                else None
+            )
+            transition_needed = next_role in {"assistant", "latest_reminder"}
+            if next_role is None:
+                transition_needed = message.task is not None or add_generation_prompt
```
[the remark this answers is not in the document]
Comment (hallerite) [reply renderers/deepseek_v4.py:716]: Confirmed. The official merge_tool_messages merges tool results into a preceding user regardless of task, while our production preprocessing incorrectly required task is None. Commit 10e421b removes that guard and adds test_tool_result_merges_into_tasked_user_like_reference_encoder, asserting token-for-token parity, task-prefix retention, and suppression of post-task reasoning. Full suite: 3659 passed, 178 skipped, 1 expected xfail.

File: renderers/deepseek_v4.py [hunk 5]
```diff
@@ -0,0 +1,809 @@
...
+                arguments = json.loads(raw_arguments)
+            except (json.JSONDecodeError, TypeError):
+                arguments = {"arguments": raw_arguments}
+        elif isinstance(raw_arguments, Mapping):
+            arguments = dict(raw_arguments)
+        else:
+            arguments = {"arguments": raw_arguments}
+
+        params: list[str] = []
+        for key, value in arguments.items():
```
[the remark this answers is not in the document]
Comment (hallerite) [reply renderers/deepseek_v4.py:358]: This is intentional exact parity with the official DeepSeek encoder, not a renderer-only failure. encode_arguments_to_dsml calls json.loads(tool_call["arguments"]) and then arguments.items(); therefore valid JSON arrays/scalars/null raise the same AttributeError upstream, while non-string values hit json.loads TypeError and are wrapped by the official except branch. Source: https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-0731/blob/main/encoding/encoding_dsv4.py#L344-L363. Signed commit f099bcb adds a parameterized regression for [], JSON string, number, and null, asserting identical exception type and message from the independent reference oracle and production renderer (23 focused tests pass). Preserving this behavior is required for the requested token/exception parity.

File: renderers/configs.py [hunk 6]
```diff
@@ -930,6 +930,43 @@ class DeepSeekR1RendererConfig(BaseRendererConfig):
...
+
+    drop_thinking: bool = True
+    """Drop reasoning before the latest user query when no tools are present.
+
+    The reference encoder automatically preserves all reasoning whenever tools
+    are supplied, regardless of this value.
+    """
+
+    reasoning_effort: Literal["low", "high", "max"] = "low"
+    """Thinking-only effort prefix; ``low`` adds no text."""
```
Comment (garrett361) [inline renderers/configs.py]: Little confusing about what the intended default is: "high" https://api-docs.deepseek.com/guides/thinking_mode/ seems to say "high" is the intended default, but I see it's `None` = low in the hf impl https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash/blob/main/encoding/encoding_dsv4.py#L222

**Review by garrett361 — COMMENTED** [review]
Thank you @hallerite ! This all looks reasonable to me, but probably someone who is more familiar w/ renderers should give the approval

File: renderers/configs.py (hunk 6 above)
Comment (hallerite) [reply renderers/configs.py]: imo the default should be low, as we are matching against the deepseek encoder, while high being the default on the API is a service-level policy

**Review by garrett361 — APPROVED** [review]
